web3.js/src/connection.ts

```diff
@@ -3877,6 +3902,31 @@ export class Connection {
     }
   }
 
+  /**
+   * Register a callback to be invoked when there is a subscribe error
+   *
+   * @param callback Function to invoke when a subscribe error occurs
+   * @returns subscription id
+   */
+  onSubscribeError(callback: SubscribeErrorCallback): number {
```

Comment:
This feels useful but I believe the main failure recently is rate limited subscriptions so I think that an error callback on each subscription method would be more valuable. I started a conversation here about this: https://github.com/solana-labs/solana/issues/19072#issuecomment-923259968

Reply:
The connection-wide open and close events sound like a good plan. Would it make sense to have Connection extend EventEmitter or should we follow a similar pattern to our existing on.*, remove.* methods?

Reply:
I think the existing methods are better because they can be strongly typed

Reply:
Okay, cool. Returning a subscription id for these seemed a little strange since they aren't directly attached to a rpc subscription, but I suppose I could refer to them as callback ids.

Reply:
Sure, that name change sounds good